### DISCUSSION TITLE
containerd: /usr/local/concourse/bin/init not found

### DISCUSSION THREAD

**Discussion by HemangLavana** [concourse/concourse#6642] [Help & Support] [answered]
I tried to enable containerd with v7.0.0 but I see the following error:

selected worker: xxx.xxx.xxx

run check: find or create container on worker xxx.xxx.xxx: starting task: new task: OCI runtime create failed: container_linux.go:370: starting container process caused: process_linux.go:459: container init caused: rootfs_linux.go:59: mounting "/usr/local/concourse/bin/init" to rootfs at "/tmp/gdn-init" caused: stat /usr/local/concourse/bin/init: no such file or directory: unknown

I am starting the worker node from CLI. I even specified CONCOURSE_CONTAINERD_INIT_BIN_PATH=/xyz/concourse/bin/init, but that didn't help either.

Not sure what am I missing.

**Comment by taylorsilva** [comment]
- How are you deploying your Concourse (helm, bosh, systemd)?
- Sanity check: if you ssh on to the worker and do `ls -l /usr/local/concourse/bin/init` what do you see?

**Comment by HemangLavana** [comment, marked as answer]
I have concourse go binaries downloaded and installed under custom path, say /xyz/concourse and not under /usr/local/concourse. Then I simply use "sudo /xyz/concourse/bin/concourse worker" to start the worker node (after I have setup few env vars for it). This approach still works if I don't enable the "export CONCOURSE_RUNTIME=containerd" before calling the "concourse worker" binary.

So, obviously /usr/local/concourse/bin/init is missing on the worker node. And I specified the env var CONCOURSE_CONTAINERD_INIT_BIN_PATH=/xyz/concourse/bin/init for it to find the script, but that doesn't help. Are there any other variables to specify a custom path or is it hard-coded?

**Reply by taylorsilva** [reply]
I think I see the problem now. The env var you're setting should not have `_PATH`. It should be `CONCOURSE_CONTAINERD_INIT_BIN`. I tried bringing up Concourse using your var and saw no change. When I removed `_PATH` it recognized the new path.

**Reply by HemangLavana** [reply]
thx, that took care of this issue. I had found this env var mentioned on https://github.com/concourse/concourse/issues/5984 but didn't realize that it's name had been changed in subsequent commits.

Now I am hitting this issue:

run check: find or create container on worker host-xxx: starting task: network add: cni net setup: failed to find plugin "bridge" in path [/usr/local/concourse/bin]

Is there another env var to specify this path?

**Reply by chenbh** [reply]
That would be `CONCOURSE_CONTAINERD_CNI_PLUGINS_DIR`.
```
          --containerd-cni-plugins-dir=                      Path to CNI network plugins. (default: /usr/local/concourse/bin) [$CONCOURSE_CONTAINERD_CNI_PLUGINS_DIR]
```

I haven't thought through the full implications of it, but maybe we should just combine those 2 into a single `CONCOURSE_BIN_DIR` that defaults to `/usr/local/concourse/bin`

**Reply by HemangLavana** [reply]
yes, to having single env var such as CONCOURSE_BIN_DIR. It seems to have gone past those issues with the new env set but I am getting a new error (see below):

selected worker: xxx-yyy-100
Cloning into '/tmp/git-resource-repo-cache'...

fatal: unable to access 'https://wwwin-github.xyz.com/ngdevx/ng-crft.git/': Could not resolve host: wwwin-github.xyz.com

The git check is trying to connect to enterpriise github instance and it seems to be failing.

**Reply by chenbh** [reply]
That sounds like a DNS problem, did you have `CONCOURSE_GARDEN_DNS_PROXY_ENABLE` set before? If so it and any other `CONCOURSE_GARDEN_*` flags can be converted to `CONCOURSE_CONTAINERD_*` flags.

The only other thing I can think of is maybe `CONCOURSE_CONTAINERD_DNS_SERVER` if Concourse failed to automatically determine servers?

**Reply by HemangLavana** [reply]
I do not have any GARDEN vars set and the moment I comment out CONTAINDERD related env vars, the same script is able to kick of worker node successfully. So I suspect there is something missing where it is unable to get DNS configs with containerd runtime. Are there any debug options or something that could be enabled to figure out what is happening.

**Reply by HemangLavana** [reply]
On further investigation, it seems the problem resolved by itself. I had left the script running and when I checked now, it seems to be working. I kicked off the worker on a different host and that also seems to be working. So I suspect that the earlier issue must have been a real network failure. 

I will close this issue now. Thanks for all your help.
